crowdsale-ico/src/lib.rs

```diff
@@ -0,0 +1,259 @@
+
+        msg::reply(IcoEvent::SaleStarted(self.ico_state.duration), 0).unwrap();
+    }
+
+    /// Purchase of tokens
+    /// 
+    /// Requirements:
+    /// * `tokens_cnt` must be greater than zero
+    /// * ICO must be in progress (already started and not finished yet)
+    /// * `msg::value` must be greater than or equal to `price * tokens_cnt`
+    /// * At least `tokens_cnt` tokens available for sale
+    /// 
+    /// Arguments:
+    /// * `tokens_cnt`: amount of tokens to purchase
+    /// 
+    pub fn buy_tokens(&mut self, tokens_cnt: u128)  {
+        let time_now: u64 = exec::block_timestamp();
+
+        assert!(tokens_cnt != 0, "buy_tokens(): Can't buy zero tokens");
+        assert!(self.ico_state.start_time + self.ico_state.duration >= time_now, "buy_tokens(): Duration of the ICO has ended");
+        assert!(self.get_balance() != 0, "buy_tokens(): All tokens have been sold");
+        self.in_process("buy_tokens()");
+
+        let current_price = self.get_current_price(time_now);
+        let cost = tokens_cnt.checked_mul(current_price)
+            .unwrap_or_else(|| panic!("buy_tokens(): Overflowing multiplication: {} * {}", tokens_cnt, current_price));
+
+        let mut change = 0;
+        let amount_sent = msg::value();
+
+        assert!(amount_sent >= cost, "buy_tokens(): Wrong amount sent, expect {} get {}", cost, amount_sent);
+        assert!(tokens_cnt <= self.get_balance(), "buy_tokens(): Not enough tokens to sell");
+
+        self.tokens_sold.checked_add(tokens_cnt)
+            .unwrap_or_else(|| panic!("buy_tokens(): Overflowing addition: {} + {}", self.tokens_sold, tokens_cnt));
+
+        if amount_sent > cost {
+            change = amount_sent - cost;
+            msg::send(msg::source(), "", change).unwrap();
+        }
+
+        self.token_holders
+            .entry(msg::source())
+            .and_modify(|balance| *balance += tokens_cnt)
+            .or_insert(tokens_cnt);
+
+        self.tokens_sold += tokens_cnt;
+        
+        msg::reply(IcoEvent::Bought { buyer: msg::source(), amount: tokens_cnt, change }, 0).unwrap();
+    }
+
+    /// Ends ICO contract
+    /// 
+    /// Requirements:
+    /// * Only owner can end ICO
+    /// * ICO can be ended more only once
+    /// * All tokens must be sold or the ICO duration must end
+    /// 
+    async fn end_sale(&mut self) {
+        let time_now: u64 = exec::block_timestamp();
+
+        assert_owner_message(&self.owner, "end_sale()");
+        self.in_process("end_sale()");
+
+        if self.ico_state.start_time + self.ico_state.duration < time_now || 
+            self.get_balance() == 0 
+        {
+            for (id, val) in &self.token_holders {
+                transfer_tokens(
+                    &self.token_id,
+                    &exec::program_id(),
+                    id,
+                    *val,
+                )
+                .await;
+            }
+        }
+
+        let rest_balance = self.get_balance();
```

Comment:
If ICO has not ended by time or by balance, will the rest balance be sent anyway? And the owner will not be able to send tokens to the holders, because `self.ico_state.ico_ended = true;` will be executed.

Reply:
And I think you need a tests for this case)